comp/metadata/inventoryagent/inventoryagent.go

```diff
@@ -0,0 +1,134 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package inventoryagent
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	flaretypes "github.com/DataDog/datadog-agent/comp/core/flare/types"
+	"github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/comp/metadata/internal/util"
+	"github.com/DataDog/datadog-agent/comp/metadata/runner"
+	"github.com/DataDog/datadog-agent/pkg/serializer"
+	"github.com/DataDog/datadog-agent/pkg/serializer/marshaler"
+	"github.com/DataDog/datadog-agent/pkg/util/hostname"
+	"go.uber.org/fx"
+)
+
+type agentMetadata map[string]interface{}
+
+// Payload handles the JSON unmarshalling of the metadata payload
+type Payload struct {
+	Hostname  string        `json:"hostname"`
+	Timestamp int64         `json:"timestamp"`
+	Metadata  agentMetadata `json:"agent_metadata"`
+}
+
+// MarshalJSON serialization a Payload to JSON
+func (p *Payload) MarshalJSON() ([]byte, error) {
+	type PayloadAlias Payload
+	return json.Marshal((*PayloadAlias)(p))
+}
+
+// SplitPayload implements marshaler.AbstractMarshaler#SplitPayload.
+//
+// In this case, the payload can't be split any further.
+func (p *Payload) SplitPayload(_ int) ([]marshaler.AbstractMarshaler, error) {
+	return nil, fmt.Errorf("could not split inventories agent payload any more, payload is too big for intake")
+}
+
+type inventoryagent struct {
+	util.InventoryPayload
+
+	log      log.Component
+	conf     config.Component
+	m        sync.Mutex
+	data     agentMetadata
+	hostname string
+}
+
+type dependencies struct {
+	fx.In
+
+	Log        log.Component
+	Config     config.Component
+	Serializer serializer.MetricSerializer
+}
+
+type provides struct {
+	fx.Out
+
+	Comp          Component
+	Provider      runner.Provider
+	FlareProvider flaretypes.Provider
+}
+
+func newInventoryAgentProvider(deps dependencies) provides {
+	hname, _ := hostname.Get(context.Background())
+	ia := &inventoryagent{
+		conf:     deps.Config,
+		log:      deps.Log,
+		hostname: hname,
+		data:     make(agentMetadata),
+	}
+
+	ia.Init(deps.Config, deps.Log, deps.Serializer, ia.getPayload, "agent.json")
+
+	if ia.Enabled {
+		ia.initData()
```

Comment:
Can you explain why this code cannot be moved inside `Init`? Should `Init` take extra parameter to handle this code?

Reply:
Not all inventory payload will initialize their data during init time. Some will fetch new data each time a payload is required.
I don't think it has a lot of value to add an extra optional parameter to `InvenotryPayload`. The goal of InventoryPayload was to encapsulate all the common logic about `Refresh`, JSON serialization, ...

Reply:
It would be nice to move `initData` in `comp/metadata/inventoryagent/inventoryagent.go`. I believed `initData` was a method of `InventoryPayload` that why I was wondering why it is not called inside `ai.Init`.